File: Src/ECScape.Core/Systems/GameStateSystem.cs
using ECScape.Components;
using ECScape.Engine;
using ECScape.Exceptions;

namespace ECScape.Systems;

internal sealed class GameStateSystem : ISystem
{
    public void Update(World world, float deltaTime)
    {
        var player = world.Entities.FirstOrDefault(x => x.HasComponent<PlayerControllable>() && x.HasComponent<Position>() && x.HasComponent<Size>());
        if (player == null) { return; }

        CheckPlayerLose(player);
        CheckPlayerWin(world.Entities);
    }

    private static void CheckPlayerLose(Entities.Entity player)
    {
        var playerSize = player.GetRequiredComponent<Size>();
        if (playerSize.Width <= 0 || playerSize.Height <= 0)
        {
            throw new GameOverException();
        }
    }

    private static void CheckPlayerWin(List<Entities.Entity> entities)
    {
        if (!entities.Where(x => x.HasComponent<Collectable>()).Any(entity => entity.GetRequiredComponent<Size>().Width > 0))
        {
            throw new GameWinException();
        }
    }
}

File: Web/src/systems/GameStateSystem.js
import GameOverException from '../core/gameOverException.js';
import ISystem from './ISystem.js';
import { Position, Size, PlayerControllable, Health } from '../components/Components.js';

class GameStateSystem extends ISystem {
    update(world, deltaTime) {
        const player = world.Entities.find(e =>
            e.hasComponent(PlayerControllable) &&
            e.hasComponent(Position) &&
            e.hasComponent(Size)
        );
        if (!player) return;

        this.checkPlayerLose(player);
    }

    checkPlayerLose(player) {
        const playerHealth = player.getRequiredComponent(Health);
        if (playerHealth.Hearts <= 0) {
            throw new GameOverException();
        }
    }
}

export default GameStateSystem;